File: crates/thicket-net/src/conn.rs
//! An authenticated channel between two resources: request/response with
//! correlation + deadlines, streaming, and an inbound queue for the serving
//! side. Built over any `AsyncRead + AsyncWrite` (in-memory duplex or TCP).

use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::time::Duration;

use snow::TransportState;
use thicket_core::{Id, WorkingKey};
use thicket_interconnect::{EnvelopePayload, EnvelopeType, SignedEnvelope};
use tokio::io::{split, AsyncRead, AsyncWrite};
use tokio::sync::{mpsc, oneshot};
use tokio::task::AbortHandle;

use crate::error::{Error, Result};
use crate::framing::{from_cbor, read_frame, to_cbor, write_frame};
use crate::identity::{LocalIdentity, VerifiedPeer};
use crate::secure;
use crate::unix_now;

type Pending = Arc<Mutex<HashMap<Vec<u8>, oneshot::Sender<SignedEnvelope>>>>;
type Streams = Arc<Mutex<HashMap<Vec<u8>, mpsc::Sender<SignedEnvelope>>>>;
type EventSubs = Arc<Mutex<HashMap<String, mpsc::Sender<SignedEnvelope>>>>;

/// How long the authentication handshake may take before the connection attempt
/// is abandoned — a silent or hostile peer must not be able to hang `connect`.
const HANDSHAKE_TIMEOUT: Duration = Duration::from_secs(10);

/// Whether a peer's handshake-validated working key is still fresh at `now`
/// (per-message freshness, plan §7). Pulled out for direct testing.
pub fn peer_key_fresh(key_not_after: u64, now: u64) -> bool {
    now <= key_not_after
}

/// An established, authenticated connection.
pub struct Conn {
    local_id: Id,
    working: WorkingKey,
    peer: VerifiedPeer,
    out_tx: mpsc::Sender<Vec<u8>>,
    pending: Pending,
    streams: Streams,
    event_subs: EventSubs,
    inbound: tokio::sync::Mutex<mpsc::Receiver<SignedEnvelope>>,
    reader_task: AbortHandle,
    writer_task: AbortHandle,
}

impl std::fmt::Debug for Conn {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Conn")
            .field("local_id", &self.local_id)
            .field("peer", &self.peer.id)
            .finish_non_exhaustive()
    }
}

impl Drop for Conn {
    fn drop(&mut self) {
        // Abort the I/O tasks so both stream halves are released promptly. Until
        // the reader half drops, a split stream stays open and the peer never
        // sees EOF — which would hang the peer's receive loop on teardown.
        self.reader_task.abort();
        self.writer_task.abort();
    }
}

impl Conn {
    /// Dial and establish an encrypted, authenticated connection (Noise
    /// initiator). If `expected_peer` is set, the authenticated peer must match.
    pub async fn connect<S>(
        stream: S,
        local: LocalIdentity,
        expected_peer: Option<Id>,
    ) -> Result<Arc<Conn>>
    where
        S: AsyncRead + AsyncWrite + Unpin + Send + 'static,
    {
        Self::establish(stream, local, expected_peer, true).await
    }

    /// Accept an incoming connection (Noise responder).
    pub async fn accept<S>(
        stream: S,
        local: LocalIdentity,
        expected_peer: Option<Id>,
    ) -> Result<Arc<Conn>>
    where
        S: AsyncRead + AsyncWrite + Unpin + Send + 'static,
    {
        Self::establish(stream, local, expected_peer, false).await
    }

    async fn establish<S>(
        mut stream: S,
        local: LocalIdentity,
        expected_peer: Option<Id>,
        initiator: bool,
    ) -> Result<Arc<Conn>>
    where
        S: AsyncRead + AsyncWrite + Unpin + Send + 'static,
    {
        let (transport, peer) = tokio::time::timeout(
            HANDSHAKE_TIMEOUT,
            secure::establish(&mut stream, &local, initiator, unix_now()),
        )
        .await
        .map_err(|_| Error::Timeout)??;
        if let Some(expected) = expected_peer {
            if expected != peer.id {
                return Err(Error::PeerMismatch);
            }
        }

        // Shared Noise transport: locked only around the synchronous AEAD op,
        // never across an await.
        let transport: Arc<Mutex<TransportState>> = Arc::new(Mutex::new(transport));
        let (mut r, mut w) = split(stream);

        let (out_tx, mut out_rx) = mpsc::channel::<Vec<u8>>(256);
        let (inbound_tx, inbound_rx) = mpsc::channel::<SignedEnvelope>(256);
        let pending: Pending = Arc::new(Mutex::new(HashMap::new()));
        let streams: Streams = Arc::new(Mutex::new(HashMap::new()));
        let event_subs: EventSubs = Arc::new(Mutex::new(HashMap::new()));

        // Writer task: encrypt each queued frame and write it to the wire.
        let transport_w = transport.clone();
        let writer_task = tokio::spawn(async move {
            while let Some(frame) = out_rx.recv().await {
                let encrypted = {
                    let mut t = transport_w.lock().unwrap();
                    secure::encrypt_frame(&mut t, &frame)
                };
                match encrypted {
                    Ok(bytes) => {
                        if write_frame(&mut w, &bytes).await.is_err() {
                            break;
                        }
                    }
                    Err(_) => break,
                }
            }
        })
        .abort_handle();

        // Reader task: decrypt and demultiplex incoming envelopes.
        let peer_r = peer.clone();
        let pending_r = pending.clone();
        let streams_r = streams.clone();
        let event_subs_r = event_subs.clone();
        let transport_r = transport.clone();
        let reader_task = tokio::spawn(async move {
            loop {
                let blob = match read_frame(&mut r).await {
                    Ok(Some(b)) => b,
                    _ => break, // EOF or error: connection done
                };
                // Per-message freshness: stop trusting the peer once its endorsed
                // working key has expired.
                if !peer_key_fresh(peer_r.key_not_after, unix_now()) {
                    break;
                }
                let plaintext = {
                    let mut t = transport_r.lock().unwrap();
                    secure::decrypt_frame(&mut t, &blob)
                };
                let plaintext = match plaintext {
                    Ok(p) => p,
                    Err(_) => break, // AEAD failure: corrupt or hostile, close
                };
                let env: SignedEnvelope = match from_cbor(&plaintext) {
                    Ok(e) => e,
                    Err(_) => continue,
                };
                // Authenticate every message against the handshake-proven key.
                if env.payload.from != peer_r.id || env.verify_with_key(&peer_r.working_pub).is_err()
                {
                    continue;
                }
                route(&env, &pending_r, &streams_r, &event_subs_r, &inbound_tx);
            }
        })
        .abort_handle();

        Ok(Arc::new(Conn {
            local_id: local.id,
            working: local.working,
            peer,
            out_tx,
            pending,
            streams,
            event_subs,
            inbound: tokio::sync::Mutex::new(inbound_rx),
            reader_task,
            writer_task,
        }))
    }

    pub fn peer(&self) -> &VerifiedPeer {
        &self.peer
    }

    pub fn local_id(&self) -> &Id {
        &self.local_id
    }

    /// Send a unary request and await its correlated response (or error), up to
    /// `timeout`. The payload is signed with the local working key.
    pub async fn call(&self, payload: EnvelopePayload, timeout: Duration) -> Result<SignedEnvelope> {
        let correlation = payload.correlation.clone();
        let signed = payload.sign(&self.working)?;

        let (tx, rx) = oneshot::channel();
        self.pending
            .lock()
            .unwrap()
            .insert(correlation.clone(), tx);

        self.out_tx
            .send(to_cbor(&signed)?)
            .await
            .map_err(|_| Error::Closed)?;

        match tokio::time::timeout(timeout, rx).await {
            Ok(Ok(env)) => Ok(env),
            Ok(Err(_)) => Err(Error::Closed),
            Err(_) => {
                self.pending.lock().unwrap().remove(&correlation);
                Err(Error::Timeout)
            }
        }
    }

    /// Send a request and receive a stream of correlated chunks. The returned
    /// receiver closes when a chunk with `stream_end` arrives or the channel ends.
    pub async fn call_stream(
        &self,
        payload: EnvelopePayload,
    ) -> Result<mpsc::Receiver<SignedEnvelope>> {
        let correlation = payload.correlation.clone();
        let signed = payload.sign(&self.working)?;

        let (tx, rx) = mpsc::channel(64);
        self.streams.lock().unwrap().insert(correlation, tx);

        self.out_tx
            .send(to_cbor(&signed)?)
            .await
            .map_err(|_| Error::Closed)?;
        Ok(rx)
    }

    /// Sign and send an envelope without awaiting a reply (responses, errors,
    /// events, stream chunks).
    pub async fn send(&self, payload: EnvelopePayload) -> Result<()> {
        let signed = payload.sign(&self.working)?;
        self.out_tx
            .send(to_cbor(&signed)?)
            .await
            .map_err(|_| Error::Closed)
    }

    /// Receive the next inbound request/event (serving side). `None` when the
    /// connection has closed.
    pub async fn recv_request(&self) -> Option<SignedEnvelope> {
        self.inbound.lock().await.recv().await
    }

    /// Subscribe to events for `topic` from the peer (plan §6 pub/sub). Returns a
    /// receiver of `Event` envelopes whose capability matches `topic`. Events for
    /// topics with no subscriber fall through to `recv_request`.
    pub fn subscribe(&self, topic: impl Into<String>) -> mpsc::Receiver<SignedEnvelope> {
        let (tx, rx) = mpsc::channel(64);
        self.event_subs.lock().unwrap().insert(topic.into(), tx);
        rx
    }

    /// Emit an event on `topic` to the peer (plan §6 `Emit`).
    pub async fn emit(&self, topic: impl Into<String>, body: Vec<u8>) -> Result<()> {
        let payload = EnvelopePayload::event(self.local_id.clone(), self.peer.id.clone(), topic)
            .with_body(body);
        self.send(payload).await
    }

    /// Best-effort flush: wait until all queued outbound frames have been handed
    /// to the writer before the connection is dropped, so a final reply is not
    /// lost to an abrupt teardown.
    pub async fn flush(&self) {
        // The writer drains `out_tx`; when capacity is fully restored the queue
        // is empty. Poll briefly rather than block forever.
        for _ in 0..1000 {
            if self.out_tx.capacity() == self.out_tx.max_capacity() {
                return;
            }
            tokio::task::yield_now().await;
        }
    }
}

/// Demultiplex one incoming envelope. Deliberately **non-blocking**: the reader
/// must never await on application consumption, or a stalled consumer on one
/// logical channel would starve response delivery on every other (head-of-line
/// deadlock). Overflow degrades gracefully — a dropped response/request makes
/// the caller time out; a dropped stream chunk truncates that one stream.
fn route(
    env: &SignedEnvelope,
    pending: &Pending,
    streams: &Streams,
    event_subs: &EventSubs,
    inbound_tx: &mpsc::Sender<SignedEnvelope>,
) {
    match env.payload.typ {
        EnvelopeType::Response | EnvelopeType::Error => {
            let waiter = pending.lock().unwrap().remove(&env.payload.correlation);
            if let Some(tx) = waiter {
                let _ = tx.send(env.clone());
            }
        }
        EnvelopeType::StreamChunk => {
            let corr = env.payload.correlation.clone();
            let mut map = streams.lock().unwrap();
            if let Some(tx) = map.get(&corr) {
                if tx.try_send(env.clone()).is_err() {
                    // Consumer stalled or gone: close (truncate) the stream.
                    map.remove(&corr);
                    return;
                }
            }
            if env.payload.stream_end {
                map.remove(&corr);
            }
        }
        EnvelopeType::Event => {
            // Deliver to a topic subscriber if one exists; otherwise surface it
            // to the serving side via the inbound queue.
            let topic = env.payload.capability.clone().unwrap_or_default();
            let sink = event_subs.lock().unwrap().get(&topic).cloned();
            match sink {
                Some(tx) => {
                    let _ = tx.try_send(env.clone());
                }
                None => {
                    let _ = inbound_tx.try_send(env.clone());
                }
            }
        }
        EnvelopeType::Request | EnvelopeType::Cancel => {
            // Drop on overflow rather than block the reader; the caller times out.
            let _ = inbound_tx.try_send(env.clone());
        }
    }
}

File: sdk/py/thicket/conn.py
"""An authenticated, encrypted connection to a peer (async).

A single **reader task** demultiplexes inbound frames by correlation: responses
and stream chunks are routed to the waiting `call` / `call_stream`; requests go to
a queue for the server side. This makes **concurrent in-flight calls on one
channel** safe — matching the Rust `Conn`. Sends are serialized because the Noise
transport assigns a sequential nonce per message.
"""

from __future__ import annotations

import asyncio
import os

from . import cbor, envelope, secure
from .framing import read_frame, write_frame
from .identity import LocalIdentity, unix_now


class Conn:
    def __init__(self, reader, writer, noise, peer, local):
        self.reader = reader
        self.writer = writer
        self.noise = noise
        self.peer = peer  # {"id": bytes, "working_pub": bytes}
        self.local = local
        self.closed_event = asyncio.Event()
        self._write_lock = asyncio.Lock()
        self._pending: dict[bytes, asyncio.Future] = {}  # correlation -> unary future
        self._streams: dict[bytes, asyncio.Queue] = {}  # correlation -> stream queue
        self._inbound: asyncio.Queue = asyncio.Queue()  # inbound requests (server side)
        self._reader_task = asyncio.ensure_future(self._reader_loop())

    # ---- establishment ----

    @classmethod
    async def connect(cls, host: str, port: int, local: LocalIdentity, expected_id=None) -> "Conn":
        reader, writer = await asyncio.open_connection(host, port)
        noise, peer = await secure.handshake(reader, writer, local, initiator=True, now=unix_now())
        if expected_id is not None and peer["id"] != expected_id:
            writer.close()
            raise ValueError("peer identity is not the expected one")
        return cls(reader, writer, noise, peer, local)

    @classmethod
    async def accept(cls, reader, writer, local: LocalIdentity, expected_id=None) -> "Conn":
        noise, peer = await secure.handshake(reader, writer, local, initiator=False, now=unix_now())
        if expected_id is not None and peer["id"] != expected_id:
            writer.close()
            raise ValueError("peer identity is not the expected one")
        return cls(reader, writer, noise, peer, local)

    # ---- reader / demux ----

    async def _reader_loop(self):
        try:
            while True:
                blob = await read_frame(self.reader)
                if blob is None:
                    break  # EOF
                try:
                    signed = cbor.decode(secure.decrypt_frame(self.noise, blob))
                except Exception:
                    break  # AEAD failure: corrupt or hostile, close
                payload = signed["payload"]
                # authenticate every message against the handshake-proven key
                if payload.get("from") != self.peer["id"]:
                    continue
                if not envelope.verify_envelope_with_key(signed, self.peer["working_pub"]):
                    continue
                self._route(signed, payload)
        finally:
            self._teardown()

    def _route(self, signed, payload):
        typ = payload.get("typ")
        corr = payload.get("correlation")
        if typ == "Request":
            self._inbound.put_nowait(signed)
        elif typ == "StreamChunk":
            q = self._streams.get(corr)
            if q is not None:
                q.put_nowait(signed)
        else:  # Response or Error — a unary reply, or a stream's terminal frame
            fut = self._pending.pop(corr, None)
            if fut is not None and not fut.done():
                fut.set_result(signed)
            else:
                q = self._streams.get(corr)
                if q is not None:
                    q.put_nowait(signed)

    def _teardown(self):
        if self.closed_event.is_set():
            return
        self.closed_event.set()
        for fut in self._pending.values():
            if not fut.done():
                fut.set_exception(ConnectionError("connection closed"))
        self._pending.clear()
        for q in self._streams.values():
            q.put_nowait(None)
        self._inbound.put_nowait(None)

    # ---- send (serialized: Noise assigns a sequential nonce) ----

    async def _send_signed(self, payload: dict) -> None:
        blob = cbor.encode(envelope.sign_envelope(payload, self.local.working))
        async with self._write_lock:
            await write_frame(self.writer, secure.encrypt_frame(self.noise, blob))

    def _request(self, capability, body, auth, context=None) -> dict:
        return envelope.build_envelope_payload(
            from_id=self.local.id,
            to_id=self.peer["id"],
            typ="Request",
            correlation=os.urandom(16),
            capability=capability,
            auth=auth,
            context=context,
            body=body,
        )

    # ---- client side ----

    async def call(self, capability: str, body: bytes = b"", *, auth=None, context=None, timeout: float = 10.0) -> dict:
        """Send a request and return the single (decoded, signed) response.
        Safe to run concurrently with other calls on the same connection."""
        if self.closed_event.is_set():
            raise ConnectionError("connection closed")
        payload = self._request(capability, body, auth, context)
        corr = payload["correlation"]
        fut = asyncio.get_running_loop().create_future()
        self._pending[corr] = fut
        try:
            await self._send_signed(payload)
            return await asyncio.wait_for(fut, timeout)
        finally:
            self._pending.pop(corr, None)

    async def call_stream(self, capability: str, body: bytes = b"", *, auth=None, context=None, timeout: float = 30.0):
        """Async-iterate response payloads until a single Response or an end-of-
        stream chunk. Safe to run concurrently with other calls."""
        if self.closed_event.is_set():
            raise ConnectionError("connection closed")
        payload = self._request(capability, body, auth, context)
        corr = payload["correlation"]
        q: asyncio.Queue = asyncio.Queue()
        self._streams[corr] = q
        try:
            await self._send_signed(payload)
            while True:
                signed = await asyncio.wait_for(q.get(), timeout)
                if signed is None:
                    return  # connection closed
                p = signed["payload"]
                if p.get("typ") == "Error":
                    raise ConnectionError((p.get("error") or {}).get("message", "remote error"))
                yield p
                if p.get("typ") == "Response" or (p.get("typ") == "StreamChunk" and p.get("stream_end")):
                    return
        finally:
            self._streams.pop(corr, None)

    def verify_response(self, signed: dict) -> bool:
        return envelope.verify_envelope_with_key(signed, self.peer["working_pub"])

    # ---- server side ----

    async def recv(self):
        """The next authenticated inbound request, or None on close. (Auth +
        from-binding are checked in the reader loop.)"""
        return await self._inbound.get()

    async def respond(self, request_payload: dict, body: bytes) -> None:
        await self._send_signed(
            envelope.build_envelope_payload(
                from_id=self.local.id,
                to_id=request_payload["from"],
                typ="Response",
                correlation=request_payload["correlation"],
                body=body,
            )
        )

    async def respond_error(self, request_payload: dict, code: str, message: str) -> None:
        await self._send_signed(
            envelope.build_envelope_payload(
                from_id=self.local.id,
                to_id=request_payload["from"],
                typ="Error",
                correlation=request_payload["correlation"],
                error={"code": code, "message": message},
            )
        )

    async def stream_chunk(self, request_payload: dict, seq: int, end: bool, body: bytes) -> None:
        await self._send_signed(
            envelope.build_envelope_payload(
                from_id=self.local.id,
                to_id=request_payload["from"],
                typ="StreamChunk",
                correlation=request_payload["correlation"],
                stream_seq=seq,
                stream_end=end,
                body=body,
            )
        )

    async def close(self) -> None:
        self._teardown()
        self._reader_task.cancel()
        self.writer.close()
        try:
            await self.writer.wait_closed()
        except Exception:
            pass
